app/services/process_state.rb

```diff
@@ -0,0 +1,45 @@
+# ProcessState is an *experimental* thing that infers the state from
+# the state of application choices. Do not rely on this for business logic.
+class ProcessState
+  def initialize(application_form)
+    @application_form = application_form
+  end
+
+  def state
+    if states.empty?
+      :unsubmitted
+    elsif unique_states == %w[unsubmitted]
+      :unsubmitted
+    elsif unique_states == %w[awaiting_references]
+      :awaiting_references
+    elsif unique_states == %w[application_complete]
+      :waiting_to_be_sent
+    elsif states.include?('awaiting_provider_decision')
+      :awaiting_provider_decisions
+    elsif states.include?('offer')
+      # Offer, but no awaiting means we're waiting on the candidate
+      :awaiting_candidate_response
+    elsif states.include?('enrolled')
+      :enrolled
+    elsif states.include?('recruited')
+      :recruited
+    elsif states.include?('pending_conditions')
+      :pending_conditions
+    else
+      :unknown_state
+    end
+  end
```

Comment:
There a few edge cases that may be slipping through the cracks here:

- Candidate has withdrawn all their application choices
- Candidate has withdrawn all (previously offered) application choices
- Providers have rejected all application choices

Reply:
Good spot! Now handled by the new `ended_without_success` state.